src/main/java/org/mitre/synthea/helpers/RandomCodeGenerator.java

```diff
@@ -26,166 +36,91 @@
    */
-  public static Code getCode(String valueSetUri, long seed) {
-    if (terminologyClient == null) {
-      throw new RuntimeException(
-          "Unable to generate code from ValueSet URI: terminology service not configured");
+  @SuppressWarnings("unchecked")
+  public static Code getCode(String valueSetUri, long seed, Code code) {
+    if (urlValidator(valueSetUri)) {
+      expandValueSet(valueSetUri);
+      List<Object> codes = codeListCache.get(valueSetUri);
+      int randomIndex = new Random(seed).nextInt(codes.size());
+      Map<String, String> codeMap = (Map<String, String>) codes.get(randomIndex);
+      validateCode(codeMap);
+      Code newCode = new Code(codeMap.get("system"), codeMap.get("code"), codeMap.get("display"));
+      selectedCodes.add(newCode);
+      return newCode;
     }
-    ValueSetExpansionComponent expansion = expandValueSet(valueSetUri);
-
-    Random random = new Random(seed);
-    int randomIndex = random.nextInt(expansion.getTotal());
-
-    ValueSetExpansionContainsComponent contains = expansion.getContains().get(randomIndex);
-    validateContains(contains);
-
-    return new Code(contains.getSystem(), contains.getCode(), contains.getDisplay());
+    return code;
   }
 
-  private static ValueSetExpansionComponent expandValueSet(String valueSetUri) {
-    ValueSet response;
-    ValueSetExpansionComponent result = new ValueSetExpansionComponent();
-    int total;
-    int offset = 0;
-    int count = 0;
+  private static boolean urlValidator(String url) {
+    if (StringUtils.isEmpty(url)) {
+      return false;
+    }
+    Matcher matcher = URL_PATTERN.matcher(url);
+    return matcher.matches();
+  }
 
-    do {
-      offset += count;
-      logger.info("Sending ValueSet expand request to terminology service (" + terminologyClient
-          .getServerBase() + "): url=" + valueSetUri + ", count=" + EXPAND_PAGE_SIZE + ", offset="
-          + offset);
+  @SuppressWarnings("unchecked")
+  private static synchronized void expandValueSet(String valueSetUri) {
+    if (!codeListCache.containsKey(valueSetUri)) {
+      HttpHeaders headers = new HttpHeaders();
+      headers.setContentType(MediaType.APPLICATION_JSON);
+      HttpEntity<String> request = new HttpEntity<>(headers);
+      Map<String, Object> valueSet = null;
       try {
-        response = responseCache.get(new ExpandInput(valueSetUri, offset));
-      } catch (ExecutionException e) {
-        throw new RuntimeException("Error expanding ValueSet", e);
+        ResponseEntity<String> response = restTemplate.exchange(expandBaseUrl + valueSetUri, HttpMethod.GET, request,
+            String.class);
+        ObjectMapper objectMapper = new ObjectMapper();
+        valueSet = objectMapper.readValue(response.getBody(), new TypeReference<Map<String, Object>>() {
+        });
+      } catch (JsonProcessingException e) {
+        throw new RuntimeException("JsonProcessingException while parsing valueSet response");
```

Comment:
Is there any consistency in the error messages that are being thrown? Usually, there's a pattern of `Error <doing something>` or `Unable to <do something>` in the errors. I scanned a few files, but didn't see one.

Consider using a more "readable" error (and use "error" instead of "exception") for this and the one on line 98. Maybe "Error parsing json while expanding ValueSet" here and "Error fetching response while expanding ValueSet" on line 98.